cloudflare/resource_cloudflare_waf_group.go

```diff
@@ -0,0 +1,203 @@
+package cloudflare
+
+import (
+	"fmt"
+	"strings"
+
+	cloudflare "github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceCloudflareWAFGroup() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareWAFGroupCreate,
+		Read:   resourceCloudflareWAFGroupRead,
+		Update: resourceCloudflareWAFGroupUpdate,
+		Delete: resourceCloudflareWAFGroupDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareWAFGroupImport,
+		},
+
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"group_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"package_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"mode": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "on",
+				ValidateFunc: validation.StringInSlice([]string{"on", "off"}, false),
+			},
+		},
+	}
+}
+
+func resourceCloudflareWAFGroupRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+
+	groupID := d.Get("group_id").(string)
+	zoneID := d.Get("zone_id").(string)
+	packageID := d.Get("package_id").(string)
+
+	group, err := client.WAFGroup(zoneID, packageID, groupID)
+	if err != nil {
+		return (err)
+	}
+
+	// Only need to set mode as that is the only attribute that could have changed
+	d.Set("mode", group.Mode)
+	d.SetId(group.ID)
+
+	return nil
+}
+
+func resourceCloudflareWAFGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	groupID := d.Get("group_id").(string)
+	zoneID := d.Get("zone_id").(string)
+	packageID := d.Get("package_id").(string)
+	mode := d.Get("mode").(string)
+
+	// If no package ID is given try to resolve it
+	var pkgList []cloudflare.WAFPackage
+	if packageID == "" {
+		var err error
+		pkgList, err = client.ListWAFPackages(zoneID)
+		if err != nil {
+			return err
+		}
+	} else {
+		pkgList = append(pkgList, cloudflare.WAFPackage{ID: packageID})
+	}
+
+	for _, pkg := range pkgList {
+		var err error
+		var group cloudflare.WAFGroup
+
+		group, err = client.WAFGroup(zoneID, pkg.ID, groupID)
+
+		if err == nil {
+			d.Set("group_id", group.ID)
```

Comment:
Not sure about the majority, as `group_id` and `package_id` and `zone_id` are expected to exist, only the `mode` is left and the call to `d.SetId`, and if there is any difference to the `mode` we want to set it to the value required by the terraform config, hence changing it ourselves and then using `d.Set` too. I'm not sure if calling `Read` would be useful here ? (seems like it would be an extra API call for nothing)

Note also that I created this file based on the logic used by `resource_cloudflare_waf_rule.go`.

Reply:
Currently, we have:

```golang
                group, err = client.WAFGroup(zoneID, pkg.ID, groupID)
                if err != nil {
                        continue
                }

                d.Set("group_id", group.ID)
                d.Set("zone_id", zoneID)
                d.Set("package_id", pkg.ID)
                d.Set("mode", mode)

                // Set the ID to the group_id parameter passed in from the user.
                // All WAF Groups already exist so we already know the group_id
                //
                // This is a work around as we are not really "creating" a WAF Group,
                // only associating it with our terraform config for future updates.
                d.SetId(group.ID)

                if group.Mode != mode {
                        err = resourceCloudflareWAFGroupUpdate(d, meta)
                        if err != nil {
                                d.SetId("")
                                return err
                        }
                }

                return nil
```

And with your suggestion, we would have:

```golang
                group, err = client.WAFGroup(zoneID, pkg.ID, groupID)
                if err != nil {
                        continue
                }

                d.Set("group_id", group.ID)
                d.Set("zone_id", zoneID)
                d.Set("package_id", pkg.ID)

                if group.Mode != mode {
                        err = resourceCloudflareWAFGroupUpdate(d, meta)
                        if err != nil {
                                d.SetId("")
                                return err
                        }
                }

                return resourceCloudflareWAFGroupRead(d, meta) // Extra call to client.WAFGroup to grab the data we just grabbed above
```

Not sure it's worth it ?

Reply:
> Not sure about the majority

There are two bits of documentation from Hashicorp on using this approach. The first is about [capturing state in `Read`](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#capture-all-state-in-read) and the second is [updating state following modifications](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#update-state-after-modification). I'd prefer we follow these best practices unless we have a pressing reason not to.

Reply:
> I'd prefer we follow these best practices unless we have a pressing reason not to.

I felt an extra API call was a good reason given there might be rate limits?
But I can change it with the code offered above.